### ISSUE TITLE
Set a color for an individual 3D vertex

### ISSUE THREAD

**Issue by kristijanbartol** [kivy/kivy#7855]
**Is your feature request related to a problem? Please describe.**
As remarkable as it sounds, after a few days and even weeks of struggle, I have been unable to set the color of a single 3D vertex of my mesh. More specifically, I have a 3D human mesh ([SMPL mesh](https://github.com/vchoutas/smplify-x)) and I would like to paint particular vertices, lines, or faces on the mesh that would describe where the particular human measurements should be taken on a real person.

**Describe the solution you'd like**
Unfortunatelly, I have not found a solution which is as simple as mesh.vertices[i].color = Color(R, G, B), not even close. I am very willing to delve into lower level Kivy code or write an "ugly" code on a higher level, but I have not found a particular file or lines of code that could possibly assign color to 3D vertex.

**Describe alternatives you've considered**
I have tried many alternatives and currently ended up with drawing individual meshes for each body measurement, which is, I hope, a temporary solution. Other alternatives that failed or are not acceptable:

1. changing the color using kivy.graphics.Color class (was simply ignored by Kivy)
2. changing fragment shader definition (only changes a global color, also can't be used to change the color after initialization)
3. providing specific texture to kivy.graphics.Mesh class (this might work if the texture is previously generated in Blender, but might be problematic to fit to people of different sizes / shapes; additionally, there is not option to possibly paint selected particular body measurements on-click or to simply highlight particular measurements more interactively)

**Additional context**
The code is an adoption of the popular ["Rotating Monkey Head" example](https://kivy.org/doc/stable/examples/gen__3Drendering__main__py.html). My rendering screen inherits from kivy.uix.screenmanager.Screen as

```
class RenderScreen(Screen):

  def __init__(self, **kwargs):
    self.canvas = kivy.graphics.RenderContext(compute_normal_mat=True)
    self.canvas.shader.source = resource_find('simple.glsl')

    super(RenderScreen, self).__init__(**kwargs)
    with self.canvas:
      # Prepare the scene and add Mesh instruction to the pipeline.
      ...
```

I create the Mesh object as:

```
self.main_mesh = Mesh(
            vertices=[],           # empty at the beginning and then generated on trigger based on the selected parameters/measurements
            indices=list(range(41328)),
            fmt=vertex_format,
            mode='triangles'
        )
```

Now I would ideally just say something like `self.main_mesh.vertices[i].color = Color (0.5, 0., 1.)`, but that seems far from possible. 

I was thinking whether the problem is in using `RenderContext` instead of `Fbo`, but concluded that this could not be the main issue. In any way, I was unable to draw a mesh like this using `Fbo`, so I sticked with the `RenderContext` for now. I am willing to push myself to make the code work with `Fbo` if `RenderContext` is controversial for some reason I did not find.

The current solution is to render multiple meshes for each body measurements:

```
self.measurement_meshes = []
        for _ in range(len(ALL_INDEX_SETS)):
            # NOTE: Indices should actually be of the expected size.
            self.measurement_meshes.append(Mesh(
                vertices=[],
                indices=list(range(495)),
                fmt=vertex_format,
                mode='triangles',
                texture=texture))
```

Rendering multiple meshes makes the application slower and it also introduces weird artifacts, as shown in the screenshot.

![github-issue-screenshot-ugly](https://user-images.githubusercontent.com/479983/161002972-92423b03-9d2f-4e68-9c21-1bb061c06986.png)

Compared to the "original", main mesh, the above render is very ugly and not presentable.

![github-issue-screenshot-original](https://user-images.githubusercontent.com/479983/161002969-dfa49847-c2f7-46cb-ab70-5e275dbf1c87.png)

**Final remarks**

While working on the prototype of this application, I always felt like I am doing something at least a bit wrong. I find it very painful to work with 3D rendering and I am never sure whether something I implemented is the proper way of doing. I was mostly relying on repositories with 3D examples such the [kivy3](https://github.com/nskrypnik/kivy3). The truth is that I am not only slightly familiar with 3D graphics and processes and it is natural that I will struggle in the beginning, but I don't see how I am going to improve when the materials and the documentation is quite scarce. I would like to improve the documentation, but first I need some real help with this supposed-to-be simple issue/feature. Thank you very much in advance!

**Comment by kristijanbartol** [comment]
Writing the issue helped me to notice a smaller problem that probably cause the artifacts:

```
self.measurement_meshes = []
        for _ in range(len(ALL_INDEX_SETS)):
            # NOTE: Indices should actually be of the expected size.
            self.measurement_meshes.append(Mesh(
                vertices=[],
                indices=list(range(495)),      # this should not be fixed to 495!!!
                fmt=vertex_format,
                mode='triangles',
                texture=texture))
```

Still, the original issue remains unsolved.

**Comment by moonpyx** [comment]
it is important that you have declared your vertex format as example binaries:

self.vertex_format = [
            (b'v_pos', 3, 'float'),
            (b'v_normal', 3, 'float'),
            (b'v_tc0', 2, 'float')]

another way is that you can make use of the ChangeState class to bind your render context and apply changes to variables in your shader

**Comment by kristijanbartol** [comment]
Thank you very much, I only now realized that the "weird" format of my vertices (converted from three to eight numbers per vertex) comes exactly from the vertex format specification. Now I have checked again what all these numbers are and see that the last two define texture coordinates for each vertex, i.e.:

```
[c0, c1, c2] -> [v0, v1, v2, n0, n1, n2, t0, t1]
```

Now I only have to figure out how to best pass the color information to these texture coordinates.

I have also previously experimented with `ChangeState` (which I forgot to mention), but probably the problem was that the `ChangeState` only affects the variables defined in the shader (as you pointed out) and if I try to do something like `ChangeState(color=(0.5, 0.5, 0.5))` and don't have `color` defined in the shader, nothing will apply. I will now try to achieve the color change using the `ChangeState` also.

Your answer greatly helped me!
